### PR TITLE
Adds support for events to use arbitrary ERC20 tokens as a deposit.

### LINKED ISSUE #236: Support any ERC20 tokens
### Prize Bounty

200 DAI
 
### Challenge Description

Currently Kickback smart contract supports any ERC20 yet we only support DAI at frontend because we hardcode lots of code.

Please refactor our code as well as extending admin page so that users can add any ERC20 tokens.

### Submission Requirements

- On our admin page, user can either chose DAI, ETH, or any ERC 20 tokens where users can copy & paste any ERX20 token address.
- It should handle different denomination (for testing purpose use 0, 8, and 18 decimal tokens which you can mint [here](https://ethereum.stackexchange.com/questions/38743/where-can-i-find-erc20-token-faucets-for-testing) )
- On our event listing page (/events), make sure that correct symbol and value are displayed.
- On our event page (/event/0x...), make sure that correct symbol and value are displayed on pot amount, RSVP amount, and the amount each participant won/withdrawn/lost.

### Submission Deadline

23rd Jan 2020
  
### Winner Announcement Date

31st Jan 2020

----------

### PR DESCRIPTION
**Pull request by TomAFrench**
## Issue number 

closes #236 
## Description 

Adds support for events to use arbitrary ERC20 tokens as a deposit.

## List of features added/changed

- Improved `getToken` query to work with contracts that both respond with their names/symbols as strings or bytes32
- Added an extra radio button labelled ERC20 to `/create` page
- Selecting ERC20 brings up an address input which only accepts amounts up to the allowed number of decimal places for that token.
- Pointed `Currency` component at `TOKEN_QUERY` to pull token symbol on rest of the UI
- Fixed misnamed props on radio buttons in `CurrencyPicker`

## How Has This Been Tested?

I've only done manual testing on the `/create` page by testing with various tokens and ensuring that they display correctly and the commitment input respects their number of decimals. I have not tested deploying a new event contract using a new token.

I've tested the display of tokens on `/events` by checking to make sure that all existing events correctly show the correct token symbol which they do with the exception of a couple of badly behaved tokens on the kovan network which don't give their symbol/name on etherscan either.

## Screenshots:

alt tag

Here shows the `/create` page on kovan with the token [Zeenus](https://kovan.etherscan.io/address/0x1f9061B953bBa0E36BF50F21876132DcF276fC6e#code), the text field disallows a decimal point to be placed. If I use [Yeenus](https://kovan.etherscan.io/address/0xc6fDe3FD2Cc2b173aEC24cc3f267cb3Cd78a26B7#code), then the text field disallows input with more than 8 digits after a decimal point.

Screenshots of the `/events` page are not shown as it is currently identical to before as no new tokens are used. 
## Checklist:

- [x] My code follows the code style of this project.
- [x] My code implements all the required features.

### CODE REVIEW

**Comment by TomAFrench** [comment]
I've been able to test this all the way through now I have kovan ETH.

This image shows an event which uses the WEENUS token as a commitment. You can see that I successfully RSVPed to the event and all values are displayed correctly

alt

**Comment by makoto** [comment]
When I tested on https://pr246-app-wearekickback.surge.sh/create, it looks like giving USD value of XEENUS  which looks like the price of ETH. Unless it's possible to lookup all token's USD value, can you disable USD value display on custom ERC 20 tokens?

Screenshot 2020-01-10 at 21 53 23

**Comment by TomAFrench** [comment]
> When I tested on https://pr246-app-wearekickback.surge.sh/create, it looks like giving USD value of XEENUS which looks like the price of ETH. Unless it's possible to lookup all token's USD value, can you disable USD value display on custom ERC 20 tokens?
> 

Ah sure thing, I don't get prices displaying when I test locally so I didn't spot that.

File: src/api/resolvers/tokenResolvers.js [hunk 1]
```diff
@@ -165,29 +309,52 @@ const resolvers = {
...
+            web3,
+            tokenAddress,
+            detailedERC20bytes32ABI
+          )
+          let [name, symbol, decimals] = await Promise.all([
+            contract.name().call(),
+            contract.symbol().call(),
+            contract.decimals().call()
+          ])
+          name = web3.utils.toAscii(name).replace(/\u0000/g, '') // eslint-disable-line no-control-regex
```
Comment (jefflau) [inline src/api/resolvers/tokenResolvers.js]: Can we please replace these with new variables to be more descriptive with what this line is doing to the variable. Same with symbol

File: src/components/SingleEvent/Admin/PartyForm.js [hunk 2]
```diff
@@ -485,35 +488,87 @@ class PartyForm extends Component {
...
+                          onChange={currencyType => {
+                            let tokenAddress
+                            if (currencyType === 'ETH') {
+                              tokenAddress = EMPTY_ADDRESS
+                            } else if (currencyType === 'DAI') {
+                              tokenAddress = this.state.daiAddress
+                            } else {
+                              tokenAddress = ''
+                            }
+                            console.log(currencyType)
```
Comment (jefflau) [inline src/components/SingleEvent/Admin/PartyForm.js]: remove log

File: src/components/SingleEvent/Admin/PartyForm.js [hunk 3]
```diff
@@ -485,35 +488,87 @@ class PartyForm extends Component {
...
-                          this.setState({ currencyType, tokenAddress })
+                      <Label>Commitment</Label>
+                      <CommitmentInput
+                        value={deposit}
+                        onChangeText={val => {
+                          let regex = /.*/
+                          if (decimals === 0 || decimals === '0') {
+                            regex = new RegExp(`^\\d*$`)
+                          } else if (decimals > 0) {
+                            regex = new RegExp(`^\\d*(\\.\\d{0,${decimals}})?$`)
```
Comment (jefflau) [inline src/components/SingleEvent/Admin/PartyForm.js]: Please rename variable to something more descriptive. And possibly add a comment if that's not enough to describe what this is doing without having to understand the regex

File: src/components/SingleEvent/Admin/PartyForm.js [hunk 4]
```diff
@@ -485,35 +488,87 @@ class PartyForm extends Component {
...
-                        daiAddress={this.state.daiAddress}
-                        onChange={({ currencyType, tokenAddress }) => {
-                          this.setState({ currencyType, tokenAddress })
+                      <Label>Commitment</Label>
+                      <CommitmentInput
+                        value={deposit}
+                        onChangeText={val => {
+                          let regex = /.*/
+                          if (decimals === 0 || decimals === '0') {
+                            regex = new RegExp(`^\\d*$`)
```
Comment (jefflau) [inline src/components/SingleEvent/Admin/PartyForm.js]: Same as comment below. Add more descriptive variable name and/or add a comment to describe

File: src/components/SingleEvent/EventCTA.js [hunk 5]
```diff
@@ -128,34 +133,56 @@ class EventCTA extends Component {
...
+                  variables={{ tokenAddress, partyAddress: address }}
+                >
+                  {({
+                    data: {
+                      tokenAllowance: { allowance, balance }
+                    },
+                    loading,
+                    refetch
+                  }) => {
+                    console.log({
```
Comment (jefflau) [inline src/components/SingleEvent/EventCTA.js]: Remove log

File: src/components/SingleEvent/EventCTA.js [hunk 6]
```diff
@@ -128,34 +133,56 @@ class EventCTA extends Component {
...
+                      decimals,
+                      allowance,
+                      balance,
+                      deposit,
+                      decoded: parseInt(toBN(deposit).toString())
+                    })
+                    console.log(
```
Comment (jefflau) [inline src/components/SingleEvent/EventCTA.js]: Remove log

File: src/routes/Faq.js [hunk 7]
```diff
@@ -55,7 +55,13 @@ class FAQ extends Component {
           invitation list.
         </p>
         <h3>Can I ask participants to commit DAI or other ERC20 tokens?</h3>
-        <p>Not yet but that's in our pipeline.</p>
+        <p>
+          Yes! We support any ERC20 token which follows the{' '}
```
Comment (jefflau) [inline src/routes/Faq.js]: Nice touch with adding to the FAQ!

Comment (TomAFrench) [reply src/routes/Faq.js]: Thanks! :smile:

File: src/components/SingleEvent/EventCTA.js (hunk 6 above)
Comment (TomAFrench) [reply src/components/SingleEvent/EventCTA.js]: done

File: src/components/SingleEvent/Admin/PartyForm.js (hunk 4 above)
Comment (TomAFrench) [reply src/components/SingleEvent/Admin/PartyForm.js]: I've added some comments and made the regex a bit more readable too

File: src/api/resolvers/tokenResolvers.js (hunk 1 above)
Comment (TomAFrench) [reply src/api/resolvers/tokenResolvers.js]: sure thing.

**Comment by TomAFrench** [comment]
Seeing as how this PR allows the use of tokens which are much less divisible than ether/dai... how does the Kickback contracts handle the case where the payout can't be split evenly? e.g. 3 zeenus being split between 2 people

**Comment by makoto** [comment]
@TomAFrench we have a `clear` function which allows organisers to clear up all the remaining dust after the cooling period

**Comment by TomAFrench** [comment]
Ah ok, that makes sense.

**Comment by makoto** [comment]
When the page is connected to wrong network, it shows this ugly error.

Screenshot 2020-01-12 at 19 18 23

@jefflau I remember we had this error before and you fixed it. Do you remember what the workaround was?

**Comment by TomAFrench** [comment]
> When the page is connected to wrong network, it shows this ugly error.
> 
> Screenshot 2020-01-12 at 19 18 23
> 
> @jefflau I remember we had this error before and you fixed it. Do you remember what the workaround was?

@makoto putting in a fix now

**Comment by TomAFrench** [comment]
What would you like to be shown in its place?

**Comment by TomAFrench** [comment]
I've done a bit of refactoring so that Currency can be given the optional prop of the amount of currency to display. It will then automatically scale it correctly and display it alongside the symbol. I've replaced all copies of Currency to make use of this.

This removed a load of manual scaling and extra queries in order to get the token decimals.

**Pull request closed without merging**